### PR TITLE
Added unique key policy support to spring data SDK

### PR DESCRIPTION
**Pull request by kushagraThapar**
# Description

* Closes https://github.com/Azure/azure-sdk-for-java/issues/15928
* Added unique key policy support to spring data SDK

* How customers will use it -> 
```
@Container
@CosmosUniqueKeyPolicy(uniqueKeys = {
    @CosmosUniqueKey(paths = {"/lastName", "/zipCode"}),
    @CosmosUniqueKey(paths = {"/city"})
})
public class CosmosUniqueKeyPolicyCodeSnippet {

    @Id
    String id;

    @PartitionKey
    String firstName;

    String lastName;
    String zipCode;
    String city;
}
```

* How the annotations look like -> 

```
/**
 * Represents the unique key policy configuration for specifying uniqueness constraints on items in the container in the
 * Azure Cosmos DB service.
 */
@Persistent
@Inherited
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface CosmosUniqueKeyPolicy {

    /**
     * Set of unique keys which enforce uniqueness constraint on items in the container
     * in the Azure Cosmos DB service.
     *
     * @return unique keys
     */
    CosmosUniqueKey[] uniqueKeys() default {};
}
```

```
/**
 * Represents a unique key on that enforces uniqueness constraint on items in the container in the Azure Cosmos DB
 * service.
 * <p>
 * 1) For containers, the value of partition key is implicitly a part of each unique key.
 * <p>
 * 2) Uniqueness constraint is also enforced for missing values.
 * <p>
 * For instance, if unique key policy defines a unique key with single property path, there could be only one item that
 * has missing value for this property.
 *
 * @see CosmosUniqueKeyPolicy
 */
@Persistent
@Inherited
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface CosmosUniqueKey {

    /**
     * A set of paths which must be unique for each item in the Azure Cosmos DB service.
     * <p>
     * The paths to enforce uniqueness on. Each path is a rooted path of the unique property in the item, such as
     * "/name/first".
     *
     * @return unique paths
     */
    String[] paths() default {};
}
```

### CODE REVIEW

**Review by milismsft — APPROVED** [review]

**Review by RaviTella — APPROVED** [review]
LGTM!

File: sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/ReactiveCosmosTemplate.java [hunk 1]
```diff
@@ -162,6 +163,10 @@ public void setApplicationContext(@NonNull ApplicationContext applicationContext
                     new CosmosContainerProperties(information.getContainerName(), information.getPartitionKeyPath());
                 cosmosContainerProperties.setDefaultTimeToLiveInSeconds(information.getTimeToLive());
                 cosmosContainerProperties.setIndexingPolicy(information.getIndexingPolicy());
+                final UniqueKeyPolicy uniqueKeyPolicy = information.getUniqueKeyPolicy();
+                if (uniqueKeyPolicy != null) {
+                    cosmosContainerProperties.setUniqueKeyPolicy(uniqueKeyPolicy);
```
[the remark this answers is not in the document]
Comment (kushagraThapar) [reply sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/ReactiveCosmosTemplate.java:168]: Just a safety check, to make sure we don't override customer's unique key policy with our empty unique key policy.

File: sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java [hunk 2]
```diff
@@ -481,6 +482,10 @@ public CosmosContainerProperties createContainerIfNotExists(CosmosEntityInformat
                     new CosmosContainerProperties(information.getContainerName(), information.getPartitionKeyPath());
                 cosmosContainerProperties.setDefaultTimeToLiveInSeconds(information.getTimeToLive());
                 cosmosContainerProperties.setIndexingPolicy(information.getIndexingPolicy());
+                final UniqueKeyPolicy uniqueKeyPolicy = information.getUniqueKeyPolicy();
+                if (uniqueKeyPolicy != null) {
+                    cosmosContainerProperties.setUniqueKeyPolicy(uniqueKeyPolicy);
```
Comment (xinlian12) [inline sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:487]: what if customer already defined some uniqueKey policy on the container? Will this override what already defined?

Comment (kushagraThapar) [reply sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:487]: If the customer already defined uniqueKey policy, then they won't use any annotation. 
Annotation is supposed to take precedence on their existing unique key policy. That's why the null check here :)

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

File: sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java (hunk 2 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:487]: If you see this PR - https://github.com/Azure/azure-sdk-for-java/pull/22246
which fixed the indexing policy override. Same intention here to add null check and only apply the indexing policy if it is not null. 
If you see `CosmosEntityInformation`, I am returning null if unique key policy is not specified or is empty.

Comment (xinlian12) [reply sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:487]: sounds good, thanks for the explanation 👍

**Pull request merged**